### ISSUE TITLE
Why making gradients zero before backward pass?

### ISSUE THREAD

**Issue by harshnova** [jcjohnson/pytorch-examples#31]
Corresponding to the line 
`model.zero_grad() ` in the
section "nn", I want to understand why do we need to make gradients zero before the backward pass?

Theoretically, isnt the backward pass supposed to utilize existing gradients to correct the weights? How setting them zero should help?

**Comment by jcjohnson** [comment]
Calling backward accumulates the gradients from the current computational graph into the gradient tensors. If you don't zero the gradients before calling backward then your gradient tensors will contain the sum of the gradients for all iterations, rather than the gradients for the current iteration; this would be an incorrect implementation of stochastic gradient descent. In practice if you forget to zero gradients before calling backward the model will sometimes train anyway, but training will be very unstable.
